keps/1618-optional-gc-of-workloads/README.md

```diff
@@ -0,0 +1,428 @@
+because "Resources" bears too many meanings in the context of Kueue 
+(Kubernetes resources, physical resources, etc.).
+- The deletion mechanism assumes that finished Workload objects do not have 
+any finalizers attached to them. After an investigation by `@mimowo`, 
+the [exact place in the code](https://github.com/kubernetes-sigs/kueue/blob/a128ae3354a56670ffa58695fef1eca270fe3a47/pkg/controller/jobframework/reconciler.go#L332-L340) 
+was found that removes the resource-in-use 
+finalizer when the Workload state transitions to finished. If that behavior ever 
+changes, or if there is an external source of a finalizer being attached 
+to the Workload, it will be marked for deletion by the Kubernetes client 
+but will not actually be deleted until the finalizer is removed.
+- If the retention policy for finished Workloads is misconfigured 
+(the provided value is not a valid time.Duration), Kueue will fail to start 
+during configuration parsing.
+- In the default scenario, where the object retention policy section 
+is not configured or finished Workloads retention is either not configured 
+or set to null, the existing behavior is maintained for backward compatibility. 
+This means that Workload objects will not be deleted by Kueue, 
+aligning with the behavior before the introduction of this feature.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- **R**: In clusters with a large number of existing finished Workloads 
+(thousands, tens of thousands, etc.), it may take a significant amount of time
+to delete all previously finished Workloads that qualify for deletion. Since the 
+reconciliation loop is effectively synchronous, this may impact the reconciliation 
+of new jobs. From the user's perspective, it may seem like Kueue's initialization 
+time is very long for that initial run until all expired objects are deleted.\
+**M**: Unfortunately, there is no easy way to mitigate this without complicating the code.
+A potential improvement would be to have a dedicated internal queue that handles deletion 
+outside the reconciliation loop, or to delete expired objects in batches so that after 
+a batch of deletions is completed, any remaining expired objects would be skipped until 
+the next run of the loop. However, this would go against the synchronous nature of the reconciliation loop.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### API
```

Comment:
TBH, I feel that this API looks like a half-bake to me.
Because when we add new policies to garbage collect other resources like integrations Jobs, we need to hard cord the API name here like `FinishedRayClusterRetention`. But, the Kueue integration Jobs are the extensible interface for external Jobs like AppWrapper or in-house one. So, I guess that it is challenging to maintain retention policies for all integrations like:

```yaml
[...]
objectRetentionPolicies:
  finiishedWorkloadRetention: 10s
  failedRayClusterRetention: 30s
  # How to specify in-house or other OSS resource policy, here?
  # In the upstream Kueue, we can not find all integrations 
  # since the integration framework could be implemented in other places regardless of private or public repositories.
[...]
```
From another perspective, even if we restrict the objective objects to specify retention duration, this API is a little bit extensible because in spite of that the API says that this API is used for retention policies for "object", but in this case, actually, we can specify only Workload retention policies.

```yaml
objectRetentionPolicies:
 finishedWorkloadRetention: 10s
 succeededWorkloadRetention: 30s 
```

In conclusion, I would ask which objects will you support in this feature through all iterations?
It's only Workload? or It's arbitrary objects?

Reply:
TBH I started working on this KEP without too much vision but rather I wanted to implement #1618 and I wanted to make the API flexible enough so that it could be extended with other object types in the future. After reading what you said it indeed makes sense to come up with a consistent approach for all objects so that user's could benefit from Kueue's extensibility. 

I don't have much experience with KEPs so I'm not sure what should happen now. For sure I'd like to ship the feature so that it doesn't slowly fades away in the in-review stage but at the same time if you feel that it's half-baked maybe I need to make a step back and come up with a more flexible approach. All feedback and guidance is really welcome 🙇 

I'll do some research over the weekend and I'll try to propose something early next week.

Reply:
There are a couple of questions / aspects I see touched here:
1. should we make it generic to cover Job CRDs?
2. should we nest the API inside `objectRetentionPolicies`?

Re 1: There is an intrinsic asymmetry between workloads and Job CRDs. Job CRDs are created by users who can specify `spec.ttlSecondsAfterFinished`. Workloads are created by Kueue, and so users don't have this control. So, I think it is justified to start from Workloads only and have the API field specific to them.

Another API maintained by Kueue that comes to my mind is ProvisioningRequest. I think having `finishedProvisioningRequestRetention` will make sense too (but we can leave it for later as currently we delete them eagerly anyway).

Re 2: I think it makes sense to nest the API inside `objectRetentionPolicies` to be future-proof in case we have some other objects with lifetime maintained by Kueue (like provisioning requests, or different policies for finished workloads depending on success  /failure).

Reply:
> Re 1: There is an intrinsic asymmetry between workloads and Job CRDs. Job CRDs are created by users who can specify spec.ttlSecondsAfterFinished. Workloads are created by Kueue, and so users don't have this control. So, I think it is justified to start from Workloads only and have the API field specific to them.

Yes, I agree with you. TBH, I don't prefer to implement the feature to do garbage collection of Jobs since I'm thinking that deletion jobs are responsible for Job's own controller like `ttlSecondsAfterFinished`. But this proposed API is extensible for Job API. So, I asked, "which objects will you support in this feature".

> Another API maintained by Kueue that comes to my mind is ProvisioningRequest. I think having finishedProvisioningRequestRetention will make sense too (but we can leave it for later as currently we delete them eagerly anyway).

I agree with you. I believe that the separate and dedicated CRDs for this feature are overkill. So, I would recommend not introducing the dedicated CRD.

> Re 2: I think it makes sense to nest the API inside objectRetentionPolicies to be future-proof in case we have some other objects with lifetime maintained by Kueue (like provisioning requests, or different policies for finished workloads depending on success /failure).

If we want to support additional Kueue-managed objects in the future,
could we add a dedicated field for each object something like this?
I guess that the criterion for garbage collection depends on the object’s kind. 

```yaml
workloadRetentionPolicy:
  onConditions:
    [metav1.Condition]
admissionCheckRetentionsPolicy:
  onConditions:
    [metav1.Condition]
  onResourceFlavor:
    name: xxx
    ...
```

Reply:
This looks a bit more complex than I originally anticipated but I agree that it's quite extensible. TBH I lack expertise when it comes evaluating whether this would be a consistent and a good way for Kueue to support that so I'll defer to other maintainers (@mimowo, @alculquicondor). Once we reach the consensus I'll be happy to implement it.

Reply:
I think that API looks too complex. For the most part, we will be happy just cleaning up finished Workloads.

Let me propose the following:

```yaml
objectRetentionPolicies:
  workloads:
    afterFinished: 10m
```

It leaves us the possibility to extend to other objects and to even add conditions, but I prefer we don't give that flexibility yet.

Reply:
> but I prefer we don't give that flexibility yet.

I understand this motivation. But, on the other hand, the proposed API still says that it can be extensible somehow in the future.

For the current API design, I am concerned that we extend the API something like below because the API doesn't restrict the target object to Workload.

```yaml
[...]
objectRetentionPolicies:
  finiishedWorkloadRetention: 10s
  acceptedAdmissionCheckRetention: 30s
[...]
```

So, I would like to select either 1. restricting the target object to Workload or 2. leaving the possibility for the other objects like this:

1

```yaml
[...]
workloadRetentionPolicies:
  afterfFinished: 10s
[...]
```

2

```yaml
objectRetentionPolicies:
  workloads:
    afterFinished: 10m
```

@alculquicondor WDYT?

Reply:
I prefer 2. It looks more future-proof.

Reply:
That sounds good to me.

Reply:
This is a side note, although. To evolve API more easily, I want to find a way to guard the new API field :)

IIUC, k/k has the feature gate guard for the new API fields like here: https://github.com/kubernetes/kubernetes/blob/1c4f540669397d031d464877ec3d2b8ac6be9b2f/pkg/apis/core/types.go#L543-L548

We can not break the new API field without bumping the API version in the current API evolving steps. I think that a new API without a guard makes API evolution harder since we need to consider future compatibility and extensibility in the first implementation.

Reply:
The feature gate only implies that the field will be ignored when disabled. It doesn't mean that we can change the API in a future version.

Reply:
Oh, I see. That was not my expected behavior.
Thanks for letting me know.